### PR TITLE
Peter/valhalla transition

### PR DESCRIPTION
**Pull request by kingofirony**
- Adds routing preview
- Uses Valhalla To receive instructions
- Pushes apk to android.mapzen.com when successfully compiled on circle
- Uses on the road 0.8

### CODE REVIEW

File: app/src/main/kotlin/util/DouglasPeuckerReducer.kt [hunk 1]
```diff
@@ -0,0 +1,148 @@
+package util
```
Comment (ecgreb) [inline app/src/main/kotlin/util/DouglasPeuckerReducer.kt]: We should move util inside com.mapzen.erasermap package

File: app/src/main/kotlin/util/HttpCacheFactory.kt [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package util
```
Comment (ecgreb) [inline app/src/main/kotlin/util/HttpCacheFactory.kt]: We should move util inside com.mapzen.erasermap package

**Comment by ecgreb** [comment]
Some asset densities are missing. All should have mdpi, hdpi, xhdpi, and xxhdpi. For example `drawable-xhdpi/ic_car.png` is not included.

File: app/src/main/kotlin/com/mapzen/erasermap/view/MainActivity.kt [hunk 3]
```diff
@@ -73,6 +87,8 @@ public class MainActivity : AppCompatActivity(), ViewController,
     var optionsMenu: Menu? = null
     var poiLayer: PoiLayer? = null
     var destination: Feature? = null
+    var path: PathLayer? = null;
```
Comment (ecgreb) [inline app/src/main/kotlin/com/mapzen/erasermap/view/MainActivity.kt]: Death to semicolons :smiling_imp:

File: app/src/main/kotlin/com/mapzen/erasermap/view/MainActivity.kt [hunk 4]
```diff
@@ -163,7 +179,8 @@ public class MainActivity : AppCompatActivity(), ViewController,
                 .setSmallestDisplacement(LOCATION_UPDATE_SMALLEST_DISPLACEMENT)
 
         LocationServices.FusedLocationApi?.requestLocationUpdates(locationRequest) {
-            location: Location ->  mapController?.showCurrentLocation(location)?.update()
+            location: Location ->
```
Comment (ecgreb) [inline app/src/main/kotlin/com/mapzen/erasermap/view/MainActivity.kt]: Why two lines?

File: app/src/main/kotlin/com/mapzen/erasermap/view/MainActivity.kt [hunk 5]
```diff
@@ -411,11 +496,61 @@ public class MainActivity : AppCompatActivity(), ViewController,
...
+        }
         presenter?.onBackPressed()
+        centerOnCurrentLocation()
     }
 
     override fun shutDown() {
         finish()
     }
+
+    private fun getInititializedRouter(): Router {
```
Comment (ecgreb) [inline app/src/main/kotlin/com/mapzen/erasermap/view/MainActivity.kt]: I think you meant `getInititititititializedRouter()` :wink:

File: app/src/main/kotlin/com/mapzen/erasermap/view/MainActivity.kt [hunk 6]
```diff
@@ -411,11 +496,61 @@ public class MainActivity : AppCompatActivity(), ViewController,
...
+                if (location is Location) {
+                    val start: DoubleArray = doubleArrayOf(location.getLatitude(), location.getLongitude())
+                    val dest: DoubleArray = doubleArrayOf(dest.getLat(), dest.getLon())
+                    getInititializedRouter().setBiking().setLocation(start).setLocation(dest).setCallback(this).fetch()
+                }
+                (findViewById(R.id.routing_circle) as ImageButton).setImageResource(R.drawable.ic_start_bike_normal)
+            }
+        }
+    }
+
```
Comment (ecgreb) [inline app/src/main/kotlin/com/mapzen/erasermap/view/MainActivity.kt]: One line between methods. Small knit I know but keeps the code tidy.

File: app/src/main/kotlin/com/mapzen/erasermap/view/MainActivity.kt [hunk 7]
```diff
@@ -399,7 +424,67 @@ public class MainActivity : AppCompatActivity(), ViewController,
...
         })
+        updateRoutePreview(destination)
+    }
+
+    fun displayRoute(route: Route?) {
+        try {
+            mapController!!.getMap().layers().remove(path)
+            mapController!!.getMap().layers().remove(markers)
+            path = PathLayer(mapController!!.getMap(), Color.DKGRAY, 8f)
+            markers = ItemizedLayer(mapController!!.getMap(), ArrayList<MarkerItem>(), AndroidGraphics.makeMarker(this.getResources().getDrawable(R.drawable.ic_pin), MarkerItem.HotspotPlace.BOTTOM_CENTER), null)
```
Comment (ecgreb) [inline app/src/main/kotlin/com/mapzen/erasermap/view/MainActivity.kt]: I haven't setup checkstyle yet (not sure if there is a Kotlin version for this?) but lines over 100 characters should be wrapped.

File: app/src/main/kotlin/com/mapzen/erasermap/view/MainActivity.kt [hunk 8]
```diff
@@ -411,11 +496,61 @@ public class MainActivity : AppCompatActivity(), ViewController,
...
+                }
+                (findViewById(R.id.routing_circle) as ImageButton).setImageResource(R.drawable.ic_start_bike_normal)
+            }
+        }
+    }
+
+
     override fun onBackPressed() {
+        if ((findViewById(R.id.route_preview)).getVisibility() == View.VISIBLE) {
+            mapController!!.getMap().layers().remove(path)
```
Comment (ecgreb) [inline app/src/main/kotlin/com/mapzen/erasermap/view/MainActivity.kt]: Better to use null safe syntax here than forced dereference. `!!` will throw NPE where `?` will skip to the next line if the value is null.

            mapController?.getMap()?.layers()?.remove(path)
            mapController?.getMap()?.layers()?.remove(markers)

File: app/src/main/kotlin/com/mapzen/erasermap/view/RoutePreviewView.kt [hunk 9]
```diff
@@ -1,24 +1,27 @@
 package com.mapzen.erasermap.view
 
 import android.content.Context
+import android.location.Location
 import android.util.AttributeSet
+import android.util.Log
 import android.view.LayoutInflater
-import android.widget.RelativeLayout
-import android.widget.TextView
+import android.widget.*
```
Comment (ecgreb) [inline app/src/main/kotlin/com/mapzen/erasermap/view/RoutePreviewView.kt]: Avoid `*` imports in favor of classes needed only.

File: app/src/main/kotlin/com/mapzen/erasermap/view/MainActivity.kt [hunk 10]
```diff
@@ -399,7 +424,67 @@ public class MainActivity : AppCompatActivity(), ViewController,
...
+
+    fun displayRoute(route: Route?) {
+        try {
+            mapController!!.getMap().layers().remove(path)
+            mapController!!.getMap().layers().remove(markers)
+            path = PathLayer(mapController!!.getMap(), Color.DKGRAY, 8f)
+            markers = ItemizedLayer(mapController!!.getMap(), ArrayList<MarkerItem>(), AndroidGraphics.makeMarker(this.getResources().getDrawable(R.drawable.ic_pin), MarkerItem.HotspotPlace.BOTTOM_CENTER), null)
+
+            var points: List<Location> = route!!.getGeometry()
+            val time = System.currentTimeMillis()
```
Comment (ecgreb) [inline app/src/main/kotlin/com/mapzen/erasermap/view/MainActivity.kt]: Unused variable

File: app/src/test/java/com/mapzen/erasermap/view/MainActivityTest.java [hunk 11]
```diff
@@ -34,17 +36,21 @@
...
 import static android.content.Context.LOCATION_SERVICE;
 import static android.view.View.GONE;
 import static android.view.View.VISIBLE;
+import static com.mapzen.erasermap.dummy.TestHelper.getFixture;
 import static com.mapzen.erasermap.dummy.TestHelper.getTestFeature;
 import static com.mapzen.erasermap.dummy.TestMap.TestAnimator.clearLastGeoPoint;
 import static com.mapzen.erasermap.dummy.TestMap.TestAnimator.getLastGeoPoint;
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.within;
+import static org.mockito.Mockito.mock;
```
Comment (ecgreb) [inline app/src/test/java/com/mapzen/erasermap/view/MainActivityTest.java]: Unused imports (3)

File: app/src/test/java/com/mapzen/erasermap/view/MainActivityTest.java [hunk 12]
```diff
@@ -360,6 +366,41 @@ public void hideRoutePreview_shouldHideRoutePreviewView() throws Exception {
...
     }
 
+
+    @Test
+    public void onSuccess_ShouldShowDrawnRoute() throws Exception {
+        activity.showRoutePreview(getTestFeature());
+        activity.success(new Route(getFixture("valhalla_route")));
+        assertThat(activity.getMapController().getMap().layers().contains(activity.getPath())).isTrue();
+    }
+
```
Comment (ecgreb) [inline app/src/test/java/com/mapzen/erasermap/view/MainActivityTest.java]: One line

File: app/src/test/java/com/mapzen/erasermap/view/MainActivityTest.java [hunk 13]
```diff
@@ -360,6 +366,41 @@ public void hideRoutePreview_shouldHideRoutePreviewView() throws Exception {
...
+
+    @Test
+    public void onBack_ShouldHideDrawnRoute() throws Exception {
```
Comment (ecgreb) [inline app/src/test/java/com/mapzen/erasermap/view/MainActivityTest.java]: Small 's' in 'should'. My bad for not setting up checkstyle.

File: app/src/main/kotlin/com/mapzen/erasermap/view/RoutePreviewView.kt [hunk 14]
```diff
@@ -1,24 +1,27 @@
...
-                + destination?.getTitle())
-    }
+        set (destination) {
+            (findViewById(R.id.destination) as TextView).setText(destination?.getTitle())
+        }
 
     public var route: Route? = null
         set (route) {
-            (findViewById(R.id.route) as TextView).setText(route?.getRawRoute().toString())
+            (findViewById(R.id.starting_point) as TextView).setText("Current Location")
```
Comment (ecgreb) [inline app/src/main/kotlin/com/mapzen/erasermap/view/RoutePreviewView.kt]: Move "Current Location" to string.xml

File: app/src/test/java/com/mapzen/erasermap/dummy/TestHelper.java [hunk 15]
```diff
@@ -42,4 +48,13 @@ public static SimpleFeature getTestSimpleFeature() {
     public static SimpleFeature getTestSimpleFeature(double lat, double lon) {
         return SimpleFeature.fromFeature(getTestFeature(lat, lon));
     }
+
```
Comment (ecgreb) [inline app/src/test/java/com/mapzen/erasermap/dummy/TestHelper.java]: One line

File: app/src/main/res/layout/view_route_preview.xml [hunk 16]
```diff
@@ -1,28 +1,64 @@
 <?xml version="1.0" encoding="utf-8"?>
-<merge xmlns:android="http://schemas.android.com/apk/res/android"
+<RelativeLayout xmlns:android="http://schemas.android.com/apk/res/android"
     android:layout_width="match_parent"
-    android:layout_height="match_parent">
+    android:layout_height="wrap_content"
+    android:layout_gravity="fill_horizontal"
+    android:background="@android:color/transparent">
 
     <!-- TODO: Replace placeholder view with the real preview header (start, destination, reverse direction, route mode, etc. -->
```
Comment (ecgreb) [inline app/src/main/res/layout/view_route_preview.xml]: TODO can be removed. You did it!

File: app/src/main/kotlin/com/mapzen/erasermap/view/MainActivity.kt (hunk 3 above)
Comment (kingofirony) [inline app/src/main/kotlin/com/mapzen/erasermap/view/MainActivity.kt]: haha old habits die hard.

File: app/src/main/kotlin/util/DouglasPeuckerReducer.kt (hunk 1 above)
Comment (kingofirony) [inline app/src/main/kotlin/util/DouglasPeuckerReducer.kt]: ✔

File: app/src/main/kotlin/util/HttpCacheFactory.kt (hunk 2 above)
Comment (kingofirony) [inline app/src/main/kotlin/util/HttpCacheFactory.kt]: ✔

File: app/src/main/kotlin/com/mapzen/erasermap/view/MainActivity.kt (hunk 3 above)
Comment (kingofirony) [inline app/src/main/kotlin/com/mapzen/erasermap/view/MainActivity.kt]: ✔ - killed the rest of them as well. really want to figure out how to write a checkstyle now

File: app/src/test/java/com/mapzen/erasermap/view/MainActivityTest.java (hunk 13 above)
Comment (kingofirony) [inline app/src/test/java/com/mapzen/erasermap/view/MainActivityTest.java]: ✔

File: app/src/test/java/com/mapzen/erasermap/view/MainActivityTest.java (hunk 12 above)
Comment (kingofirony) [inline app/src/test/java/com/mapzen/erasermap/view/MainActivityTest.java]: ✔

File: app/src/test/java/com/mapzen/erasermap/view/MainActivityTest.java (hunk 11 above)
Comment (kingofirony) [inline app/src/test/java/com/mapzen/erasermap/view/MainActivityTest.java]: ✔

File: app/src/main/kotlin/com/mapzen/erasermap/view/MainActivity.kt (hunk 10 above)
Comment (kingofirony) [inline app/src/main/kotlin/com/mapzen/erasermap/view/MainActivity.kt]: ✔

File: app/src/main/kotlin/util/DouglasPeuckerReducer.kt (hunk 1 above)
Comment (kingofirony) [inline app/src/main/kotlin/util/DouglasPeuckerReducer.kt]: ✔

File: app/src/main/kotlin/util/HttpCacheFactory.kt (hunk 2 above)
Comment (kingofirony) [inline app/src/main/kotlin/util/HttpCacheFactory.kt]: ✔

File: app/src/main/kotlin/com/mapzen/erasermap/view/MainActivity.kt (hunk 4 above)
Comment (kingofirony) [inline app/src/main/kotlin/com/mapzen/erasermap/view/MainActivity.kt]: typo i guess - ✔

File: app/src/main/kotlin/com/mapzen/erasermap/view/MainActivity.kt (hunk 5 above)
Comment (kingofirony) [inline app/src/main/kotlin/com/mapzen/erasermap/view/MainActivity.kt]: HAHA. 
✔

File: app/src/main/kotlin/com/mapzen/erasermap/view/MainActivity.kt (hunk 6 above)
Comment (kingofirony) [inline app/src/main/kotlin/com/mapzen/erasermap/view/MainActivity.kt]: ✔

File: app/src/main/kotlin/com/mapzen/erasermap/view/MainActivity.kt (hunk 7 above)
Comment (kingofirony) [inline app/src/main/kotlin/com/mapzen/erasermap/view/MainActivity.kt]: ✔

File: app/src/main/kotlin/com/mapzen/erasermap/view/RoutePreviewView.kt (hunk 9 above)
Comment (kingofirony) [inline app/src/main/kotlin/com/mapzen/erasermap/view/RoutePreviewView.kt]: ✔

File: app/src/main/kotlin/com/mapzen/erasermap/view/RoutePreviewView.kt (hunk 14 above)
Comment (kingofirony) [inline app/src/main/kotlin/com/mapzen/erasermap/view/RoutePreviewView.kt]: ✔

File: app/src/test/java/com/mapzen/erasermap/dummy/TestHelper.java (hunk 15 above)
Comment (kingofirony) [inline app/src/test/java/com/mapzen/erasermap/dummy/TestHelper.java]: ✔

File: app/src/main/res/layout/view_route_preview.xml (hunk 16 above)
Comment (kingofirony) [inline app/src/main/res/layout/view_route_preview.xml]: ✔

File: app/src/main/kotlin/com/mapzen/erasermap/view/MainActivity.kt (hunk 8 above)
Comment (kingofirony) [inline app/src/main/kotlin/com/mapzen/erasermap/view/MainActivity.kt]: ✔

**Pull request merged**